doc: add note to ninjia build for macOS using -jn flag (#53187)
nodejs/node: doc/contributing/building-node-with-ninja.md

         # regardless of the number of cores on the current machine.
```

Note: if you are on macOS and use GNU Make version `3.x`, the `-jn` flag
will not work. You can either upgrade to `v4.x` (e.g. using a package manager

> such as [Homebrew](https://formulae.brew.sh/formula/make#default)) or use `make JOBS=n`.

Reviewer: If `JOBS=n` works for all `make` versions, why don’t we just change the docs to recommend it as the only solution? It’s much more straightforward than telling people to upgrade their `make`.

Author: IMO I guess it would be better to leave people with options that they can choose? e.g. some people have already used to the `make -jn` and wanna keep using it across Linux and Mac.

Reviewer: What's the advantage of listing both options if they do the same thing?

Reviewer 2: `make -jn` is a widely used convention and shorter to type than setting `JOBS`. Also, if one accidentally mistyped `-jn`, there will almost certainly be an error, whereas mistyping the name of a variable usually doesn't even give a warning.

I don't know why Apple ships an ancient version of `make` by default, but it seems that upgrading it is easy enough.

Reviewer: I upgraded it via brew and I still got the error. Switching to the variable worked for me. Presumably the upgrade didn't put the new version in my path or something but I didn't dig into it.

Author: Hi @GeoffreyBooth, I think by default `homebrew` won't put it in your path.

> GNU "make" has been installed as "gmake". If you need to use it as "make", you can add a "gnubin" directoryto your PATH from your bashrc like: PATH="$HOMEBREW_PREFIX/opt/make/libexec/gnubin:$PATH" [^gmake]

[^gmake]: https://formulae.brew.sh/formula/make

Reviewer: So essentially the instructions are:

> Use `JOBS=N`; or if you’re on macOS and you want to type `-jN` instead, use Homebrew to upgrade `make` and add `$HOMEBREW_PREFIX/opt/make/libexec/gnubin` to your `$PATH`

The second option is so complex that I don’t really see the point in mentioning it?

Reviewer 2: Honestly it's up to Apple to provide their users with a version of `make` that isn't old enough to vote in most countries.

As long as they don't, we can — and probably should — give users both options. Clearly, there is an expectation among contributors that `-jn` should work, and it can.

If you think the details of upgrading to a version of `make` from the last decade are too complex/verbose, we can probably omit those and just say "please upgrade to version 4 of GNU make" as option two.

In any case, if probably doesn't hurt to make `does not work` a hyperlink to the issue that Yagiz created so that readers can find plenty of context there.

Reviewer: I also don’t understand how this broke. I’ve used `-j8` for years on Mac and it was only recently that it started erroring for me, and fortunately I happened to find this PR. So I’m all for updating the instructions, but did we change something recently that requires a newer `make`? If so, is this the _only_ thing that’s broken in the old `make` or should we do some kind of check that users are running `make` 4+?

If using the old `make` is okay, I would update all the places we currently mention `-jN` to instead be `JOBS=N` and then add a note like “If you are using a current version of `make`, which is likely for most Linux OSes but currently not for macOS, you can replace `JOBS=N` with `-jN`.” As in, we should default to the instruction that should work for everyone.

Reviewer 3: FWIW I installed an up-to-date version using Nix, I had to restart my shell and it worked fine. I'm -1 on Goeffrey's proposal, I agree with Tobias it's up to the user to use an up-to-date version of `make`. If it's upgrading is too complicated, not passing `-j` is also an option – and you'd get the same behavior as before – and also building with ninja is not the default.

Author: > I also don’t understand how this broke. I’ve used -j8 for years on Mac and it was only recently that it started erroring for me

The sad story is that `-jn` may never have truly worked. see https://github.com/nodejs/node/issues/53176#issuecomment-2134759593,
https://github.com/nodejs/node/issues/53176#issuecomment-2134780877,
https://github.com/nodejs/node/issues/53176#issuecomment-2134799965.

(the tldr is that I have tested on the previous working version and found the `-jn` was not being propagated)

         # regardless of the number of cores on the current machine.
```

> Note: if you are on macOS and use GNU Make version `v3.8.x`, the `-jn` flag

Reviewer 2: It's `3.81` (not `3.8.1`), but I think it's fine to broaden that to `3.x`:

```suggestion
Note: if you are on macOS and use GNU Make version `3.x`, the `-jn` flag
```

Also, the `v` is redundant since you explicitly wrote `version` before that.

> will be ignored, you can either upgrade to v4.x from [Homebrew](https://formulae.brew.sh/formula/make#default) or use `make JOBS=n`.

Reviewer 2: Based on https://github.com/nodejs/node/issues/53176, it's not ignored though but will cause an error:

```suggestion
will not work. You can either upgrade to v4.x from [Homebrew](https://formulae.brew.sh/formula/make#default) or use `make JOBS=n`.
```

Reviewer 3: ```suggestion
will not work. You can either upgrade to v4.x (e.g. using a package manager
such as [Homebrew](https://formulae.brew.sh/formula/make#default)) or use `make JOBS=n`.
```

Author: thanks @tniessen @aduh95 🙏
